### PR TITLE
clean ast to use more standard encoding

### PR DESCRIPTION
**Pull request by MikeHolman**
There were a number of places in ast which we were not decoding in same was as spidermonkey, now most of these inconsistencies should be resolved.

---
This change is [Review on Reviewable](https://reviewable.io/reviews/microsoft/chakracore/492)

### CODE REVIEW

File: lib/WasmReader/WasmBinaryReader.cpp [hunk 1]
```diff
@@ -330,6 +331,7 @@ WasmBinaryReader::ASTNode()
         BlockNode();
         break;
     case wbCall:
+    case wbCallImport:
```
Comment (MikeHolman) [inline lib/WasmReader/WasmBinaryReader.cpp]: I added this as tmp measure, but will need some convergence with import table change

**Comment by MikeHolman** [comment]
@commonlisp @Cellule

File: lib/WasmReader/WasmBinaryReader.cpp [hunk 2]
```diff
@@ -616,40 +611,25 @@ WasmBinaryReader::FunctionBodyHeader()
...
         {
         case Wasm::WasmTypes::I32:
-            i32Count += count;
+            m_funcInfo->AddLocal(Wasm::WasmTypes::I32, count);
+            TRACE_WASM_DECODER(L"Function body header: I32 local");
             break;
         case Wasm::WasmTypes::I64:
-            i64Count += count;
+            m_funcInfo->AddLocal(Wasm::WasmTypes::I64, count);
+            TRACE_WASM_DECODER(L"Function body header: I64 local");
```
Comment (MikeHolman) [inline lib/WasmReader/WasmBinaryReader.cpp]: Looking now, this whole switch is kind of pointless now... Should just AddLocal(type, count). Will update in morning with this.

File: lib/WasmReader/WasmByteCodeGenerator.cpp [hunk 3]
```diff
@@ -482,8 +514,11 @@ WasmBytecodeGenerator::EmitLoop()
         {
             op = m_reader->ReadFromBlock();
             EmitInfo info = EmitExpr(op);
+            GetRegisterSpace(info.type)->ReleaseLocation(&info);
```
Comment (MikeHolman) [inline lib/WasmReader/WasmByteCodeGenerator.cpp]: I should check against Void type here

File: lib/WasmReader/WasmByteCodeGenerator.cpp [hunk 4]
```diff
@@ -774,35 +810,50 @@ WasmBytecodeGenerator::EmitIfElseExpr()
...
-    // we will use result info true expr as our result info
-    // review: how will this work with break? must we store this location?
-    if (falseExpr.type != WasmTypes::Void && falseExpr.location != trueExpr.location)
+    else
     {
-        Js::OpCodeAsmJs op = Js::OpCodeAsmJs::Nop;
-        switch (falseExpr.type)
+        // we will use result info true expr as our result info
+        // review: how will this work with break? must we store this location?
+        if (falseExpr.type != WasmTypes::Void && falseExpr.location != trueExpr.location)
```
Comment (MikeHolman) [inline lib/WasmReader/WasmByteCodeGenerator.cpp]: Can remove check against void

**Comment by Cellule** [comment]
Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[lib/WasmReader/WasmBinaryReader.cpp, line 413 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRjDlqpCr1Ci1hUbPy:-KCRjDlqpCr1Ci1hUbPz:-263388655) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmBinaryReader.cpp#L413)):*
Should update the decoding error

---

*[lib/WasmReader/WasmBinaryReader.cpp, line 619 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCP5D3E6piljd8to90w-r1-619:-KCRkA73pkeTktwqS29w:-991739440) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmBinaryReader.cpp#L619)):*
If you do remove the switch case, make sure to check for `type < WasmTypes::Limit`

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 296 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRn2DPpyTpVKEyKMOm:-KCRn2DPpyTpVKEyKMOn:-1241430263) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L296)):*
Here you can do
```cpp
#define MakeWide(x) L##x
    switch (op)
    {
#define WASM_KEYWORD(token, name) case wn##token: Output::Print(MakeWide(#token ## "\r\n")); break;
#include "WasmKeywords.h"
    }
#undef MakeWide
```
Technically we already have a `MakeWide` macro `_u`, but only on the master branch at this time

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 311 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRoyCF1mp4u518CgLX:-KCRoyCF1mp4u518CgLY:-930944041) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L311)):*
didn't you use the ByteCode phase ?

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 399 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRp8Tu2UqajVT69eVg:-KCRp8Tu2UqajVT69eVh:120863613) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L399)):*
`<=` ?
I find `if (localNum >= m_funcInfo->GetLocalCount())` easier to read, but that could just be me.

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 420 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRqqwDncXeBJQPDjqk:-KCRqqwDncXeBJQPDjql:121421134) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L420)):*
Looking at https://github.com/WebAssembly/design/blob/master/AstSemantics.md#order-of-evaluation
I believe the tmp register is not necessary. 
It is hard to tell if there could be side-effects, but I doubt it.

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 517 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCP5D3E6piljd8to90y-r1-517:-KCRrKrLdTBibumS_7y-:998995845) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L517)):*
I think it would be easier to have a dummy void register space that no-op operations.

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 541 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRr_lTcOCoZN1W3c7z:-KCRr_lTcOCoZN1W3c8-:492571941) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L541)):*
This is no longer a forward call it's a decoder error since we should have all functions at this point

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 987 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRryANPKOXJvNIK4Bo:-KCRryAOLm3uH6IMMUsk:-344023341) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L987)):*
why the change here ?

---

*[lib/WasmReader/WasmSections.h, line 10 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRl0hBVO__Ttv4CjXE:-KCRl0hBVO__Ttv4CjXF:1685442598) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmSections.h#L10)):*
Do not commit those changes please

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/492)*

**Comment by MikeHolman** [comment]
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[lib/WasmReader/WasmBinaryReader.cpp, line 413 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRjDlqpCr1Ci1hUbPy:-KCS4PHacvdPQhZ4rKQY:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmBinaryReader.cpp#L413)):*
Done.

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 296 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRn2DPpyTpVKEyKMOm:-KCS4u5QGcwwWAwAdKxA:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L296)):*
Done.

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 311 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRoyCF1mp4u518CgLX:-KCRxBzjgBaKZIsYwPRK:550130744) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L311)):*
changed my  mind. I can switch if you would prefer not to have that spew for WasmReader phase.

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 399 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRp8Tu2UqajVT69eVg:-KCSNqh3szeRSqgjz3uw:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L399)):*
Done.

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 420 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRqqwDncXeBJQPDjqk:-KCRzBq_TQngqrfv4h29:-1635114544) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L420)):*
looks like it is necessary to move to a temp. We can think about optimizing later in case the result isn't used.

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 517 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCP5D3E6piljd8to90y-r1-517:-KCRzL9VK9hZefIO2feX:-30261829) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L517)):*
I made a helper to do this. Let me know what you think about it.

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 987 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRryANPKOXJvNIK4Bo:-KCS3Wvp5sGjdXbHvcoM:958930723) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L987)):*
was changing the return since it now gives int and noticed we can directly use m_i32RegSlots without needing to GetRegSpace(), which I stylistically prefer.

---

*[lib/WasmReader/WasmSections.h, line 10 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRl0hBVO__Ttv4CjXE:-KCS4JK9PytBRYl3rHwT:-1816433417) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmSections.h#L10)):*
whoops, forgot about that!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/492)*

**Comment by Cellule** [comment]
Reviewed 4 of 4 files at r2, 8 of 8 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[lib/WasmReader/WasmBinaryReader.cpp, line 413 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRjDlqpCr1Ci1hUbPy:-KCSRrAF4jbKKHVXxREI:29384506) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmBinaryReader.cpp#L413)):*
I don't see any changes here

---

*[lib/WasmReader/WasmByteCodeGenerator.cpp, line 311 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRoyCF1mp4u518CgLX:-KCSSFy1Mf9wSoro_TIq:617963683) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmByteCodeGenerator.cpp#L311)):*
It doesn't really matter at this point

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/492)*

**Comment by Cellule** [comment]
LGTM :+1:

**Comment by MikeHolman** [comment]
*[lib/WasmReader/WasmBinaryReader.cpp, line 413 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/492#-KCRjDlqpCr1Ci1hUbPy:-KCSXVGJ9sJMo-nKpJW2:-967364410) ([raw file](https://github.com/microsoft/chakracore/blob/83057e365e0775b26f566a2655278aea5f0ec048/lib/WasmReader/WasmBinaryReader.cpp#L413)):*
oops, now done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/microsoft/chakracore/492)*

**Pull request merged**